src/index.ts

```diff
@@ -121,45 +121,85 @@ export interface ZodClass<
 }
 
 type OptionalKeys<Shape> = {
-  [k in keyof Shape]: undefined extends Z.infer<Shape[k]> ? k : never;
+  [k in keyof Shape]: undefined extends Z.output<Shape[k]> ? k : never;
 }[keyof Shape];
 
 export declare namespace Z {
-  export type infer<T> = T extends new (...args: any[]) => infer R
+  export type output<T> = T extends new (...args: any[]) => infer R
     ? R
     : T extends ZodObject<infer Shape>
     ? {
         [k in keyof Pick<
           Shape,
           Exclude<keyof Shape, OptionalKeys<Shape>>
-        >]: Z.infer<Shape[k]>;
+        >]: Z.output<Shape[k]>;
       } & {
-        [k in OptionalKeys<Shape>]+?: Z.infer<Shape[k]>;
+        [k in OptionalKeys<Shape>]+?: Z.output<Shape[k]>;
       }
     : T extends ZodArray<infer I>
-    ? Z.infer<I>[]
+    ? Z.output<I>[]
     : T extends ZodOptional<infer T>
-    ? Z.infer<T> | undefined
+    ? Z.output<T> | undefined
     : T extends ZodNullable<infer T>
-    ? Z.infer<T> | null
+    ? Z.output<T> | null
     : T extends ZodTuple<infer T>
-    ? { [i in keyof T]: Z.infer<T[i]> }
+    ? { [i in keyof T]: Z.output<T[i]> }
     : T extends ZodRecord<infer Key, infer Value>
     ? {
-        [k in Extract<Z.infer<Key>, string | number | symbol>]: Z.infer<Value>;
+        [k in Extract<
+          Z.output<Key>,
+          string | number | symbol
+        >]: Z.output<Value>;
       }
     : T extends ZodMap<infer Key, infer Value>
-    ? Map<Z.infer<Key>, Z.infer<Value>>
+    ? Map<Z.output<Key>, Z.output<Value>>
     : T extends ZodSet<infer Item>
-    ? Set<Z.infer<Item>>
+    ? Set<Z.output<Item>>
     : T extends ZodFunction<infer Args, infer Output>
-    ? (...args: Z.infer<Args>) => Z.infer<Output>
+    ? (...args: Z.output<Args>) => Z.output<Output>
     : T extends ZodLazy<infer T>
-    ? Z.infer<T>
+    ? Z.output<T>
     : T extends ZodPromise<infer T>
-    ? Promise<Z.infer<T>>
+    ? Promise<Z.output<T>>
     : T extends ZodType<any, any, any>
-    ? z.infer<T>
+    ? z.output<T>
+    : never;
+
+  export type input<T> = T extends new (...args: any[]) => infer R
```

Comment:
Curious, what was the reason to separate out into input/output?

Reply:
The constructors argument should be of the type `z.input<schema>` and the instance type of the class should be `z.output<schema>` with `z.input` being the type of what should go into `z.parse` and `z.output` is the type it returns if parsed successfully which includes all transforms which can change the type.

I could try to parameterize this type and have `z.input` and `z.output` passed in as a type parameter to reduce the duplication. Should I give that a try?

I hope I got the question right, if not let me know :)